File: samples/bot-message-reaction/csharp/MessageReaction/Log/ActivityLog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;

namespace MessageReaction.Log
{
    /// <summary>
    /// Manages the logging of activities.
    /// </summary>
    public class ActivityLog
    {
        private readonly IStorage _storage;

        /// <summary>
        /// Initializes a new instance of the <see cref="ActivityLog"/> class.
        /// </summary>
        /// <param name="storage">The storage to use for logging activities.</param>
        public ActivityLog(IStorage storage)
        {
            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
        }

        /// <summary>
        /// Appends an activity to the log.
        /// </summary>
        /// <param name="activityId">The ID of the activity.</param>
        /// <param name="activity">The activity to log.</param>
        /// <returns>A task that represents the asynchronous operation.</returns>
        /// <exception cref="ArgumentNullException">Thrown when activityId or activity is null.</exception>
        public async Task AppendAsync(string activityId, Activity activity)
        {
            if (string.IsNullOrEmpty(activityId))
            {
                throw new ArgumentNullException(nameof(activityId));
            }

            if (activity == null)
            {
                throw new ArgumentNullException(nameof(activity));
            }

            var data = new Dictionary<string, object> { { activityId, activity } };
            await _storage.WriteAsync(data);
        }

        /// <summary>
        /// Finds an activity in the log by its ID.
        /// </summary>
        /// <param name="activityId">The ID of the activity to find.</param>
        /// <returns>The activity if found; otherwise, null.</returns>
        /// <exception cref="ArgumentNullException">Thrown when activityId is null.</exception>
        public async Task<Activity> FindAsync(string activityId)
        {
            if (string.IsNullOrEmpty(activityId))
            {
                throw new ArgumentNullException(nameof(activityId));
            }

            var activities = await _storage.ReadAsync(new[] { activityId });
            return activities.TryGetValue(activityId, out var activity) ? (Activity)activity : null;
        }
    }
}

File: samples/bot-message-reaction/nodejs/activityLog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

/**
 * ActivityLog handles the storage and retrieval of activities.
 * This is used to track messages sent by the bot so we can respond to reactions.
 * @class
 */
class ActivityLog {
    /**
     * Creates an instance of ActivityLog.
     * @param {Object} storage - The storage mechanism for activities.
     */
    constructor(storage) {
        this._storage = storage;
    }

    /**
     * Appends an activity to the log with its activity ID.
     * @param {string} activityId - The ID of the activity.
     * @param {Object} activity - The activity to append.
     * @throws {TypeError} If activityId or activity is null.
     */
    async append(activityId, activity) {
        if (activityId == null) {
            throw new TypeError('activityId is required for ActivityLog.append');
        }
        
        if (activity == null) {
            throw new TypeError('activity is required for ActivityLog.append');
        }

        const conversationId = activity.conversation?.id || 'unknown';
        const key = `activities_${conversationId}`;
        
        // Get existing activities for this conversation
        let activities = this._storage.get(key) || {};
        
        // Store the activity with its ID
        activities[activityId] = { 
            activity: {
                text: activity.text,
                type: activity.type,
                timestamp: new Date().toISOString()
            }
        };
        
        // Save back to storage
        this._storage.set(key, activities);
    }

    /**
     * Finds an activity in the log by its ID.
     * @param {string} activityId - The ID of the activity to find.
     * @param {string} conversationId - The conversation ID to search in.
     * @returns {Object|null} The activity if found, otherwise null.
     * @throws {TypeError} If activityId is null.
     */
    async find(activityId, conversationId = 'unknown') {
        if (activityId == null) {
            throw new TypeError('activityId is required for ActivityLog.find');
        }

        const key = `activities_${conversationId}`;
        const activities = this._storage.get(key) || {};
        
        return activities[activityId] ? activities[activityId].activity : null;
    }

    /**
     * Clears all activities for a conversation.
     * @param {string} conversationId - The conversation ID to clear.
     */
    async clear(conversationId) {
        const key = `activities_${conversationId}`;
        this._storage.delete(key);
    }
}

module.exports = { ActivityLog };